File: tests/test_ui_chaos.py
import pytest
import os
import random
from playwright.async_api import Page, expect


# Minimal navigation fixture relying on pytest-playwright to handle loop
@pytest.fixture(autouse=True)
async def navigate_to_app(page: Page):
    """Navigate to app."""
    port = os.environ.get("NICEGUI_SCREEN_TEST_PORT", "8080")
    url = f"http://localhost:{port}"
    try:
        await page.goto(url, timeout=10000)
    except Exception as e:
        pytest.skip(f"App not running at {url}: {e}")
    yield


@pytest.mark.asyncio
def _do_test_chaos_monkey_setup():
    """Helper: setup phase for test_chaos_monkey."""

    print("\n🐒 Unleashing Chaos Monkey...")

    selectors = ["button", "input[class*='q-field__native']", ".q-toggle", ".q-btn"]

    interactions = 25
    errors = []

    return errors, interactions, selectors


async def test_chaos_monkey(page: Page):
    """Test chaos monkey."""
    errors, interactions, selectors = _do_test_chaos_monkey_setup()
    # Wait for app to be ready
    await page.wait_for_timeout(2000)

    for i in range(interactions):
        try:
            # Re-scan DOM each time
            candidates = []
            for sel in selectors:
                elements = await page.locator(sel).all()
                for el in elements:
                    if await el.is_visible():
                        candidates.append(el)

            if not candidates:
                await page.wait_for_timeout(200)
                continue

            # Pick random victim
            target = random.choice(candidates)

            try:
                tag_name = await target.evaluate("el => el.tagName.toLowerCase()", timeout=200)

                if tag_name == "input":
                    text = "".join(random.choices("abcdef", k=3))
                    await target.fill(text, timeout=200)
                else:
                    await target.click(timeout=200, no_wait_after=True)
            except Exception:
                pass

            # Check for crash
            if await page.locator("text='Connection lost'").is_visible():
                errors.append(f"Connection Lost detected at step {i}")
                break

            await page.wait_for_timeout(50)

        except Exception:
            pass

    if errors:
        pytest.fail(f"❌ Application Crashed: {errors}")

    assert True
    print("✅ Application Survived!")

File: rust_output/src/test_ui_chaos.rs
use anyhow::{Result, Context};
use tokio;

/// Navigate to app.
pub async fn navigate_to_app(page: Page) -> Result<()> {
    // Navigate to app.
    let mut port = std::env::var(&"NICEGUI_SCREEN_TEST_PORT".to_string()).unwrap_or_default().cloned().unwrap_or("8080".to_string());
    let mut url = format!("http://localhost:{}", port);
    // try:
    {
        page.goto(url, /* timeout= */ 10000).await;
    }
    // except Exception as e:
    Ok(/* yield */)
}

/// Helper: setup phase for test_chaos_monkey.
pub fn _do_test_chaos_monkey_setup() -> () {
    // Helper: setup phase for test_chaos_monkey.
    println!("{}", "\n🐒 Unleashing Chaos Monkey...".to_string());
    let mut selectors = vec!["button".to_string(), "input[class*='q-field__native']".to_string(), ".q-toggle".to_string(), ".q-btn".to_string()];
    let mut interactions = 25;
    let mut errors = vec![];
    (errors, interactions, selectors)
}

/// Test chaos monkey.
pub async fn test_chaos_monkey(page: Page) -> Result<()> {
    // Test chaos monkey.
    let (mut errors, mut interactions, mut selectors) = _do_test_chaos_monkey_setup();
    page.wait_for_timeout(2000).await;
    for i in 0..interactions.iter() {
        // try:
        {
            let mut candidates = vec![];
            for sel in selectors.iter() {
                let mut elements = page.locator(sel).all().await;
                for el in elements.iter() {
                    if el.is_visible().await {
                        candidates.push(el);
                    }
                }
            }
            if !candidates {
                page.wait_for_timeout(200).await;
                continue;
            }
            let mut target = random.choice(candidates);
            // try:
            {
                let mut tag_name = target.evaluate("el => el.tagName.toLowerCase()".to_string(), /* timeout= */ 200).await;
                if tag_name == "input".to_string() {
                    let mut text = random.choices("abcdef".to_string(), /* k= */ 3).join(&"".to_string());
                    target.fill(text, /* timeout= */ 200).await;
                } else {
                    target.click(/* timeout= */ 200, /* no_wait_after= */ true).await;
                }
            }
            // except Exception as _e:
            if page.locator("text='Connection lost'".to_string()).is_visible().await {
                errors.push(format!("Connection Lost detected at step {}", i));
                break;
            }
            page.wait_for_timeout(50).await;
        }
        // except Exception as _e:
    }
    if errors {
        pytest.fail(format!("❌ Application Crashed: {}", errors));
    }
    assert!(true);
    Ok(println!("{}", "✅ Application Survived!".to_string()))
}
